LFS-655: Handle re-login globally

IntegralProgrammer opened the pull request:
This PR is in _draft mode_ as the standardized re-login procedure has not been added to all `fetch()` calls requiring authentication. Nonetheless, any feedback on the changed functionalities thus far would be greatly appreciated. These changed functionalities are:

- _User dashboard_
- _Subjects_ page
- _Forms_ page

- Any individual subject page (eg. a patient chart)
- Any individual form (eg. a Demographics Data form)

- The Delete button (eg. delete a form), has some minor issues handling incorrect logins when deleting a form when the form is open.

To test, login to LFS in two different browser tabs. On one browser tab logout. On the other tab, try the various above listed functionalities (eg. visit the _User dashboard_, visit the _Subjects page_, etc...)

marta- reviewed (commented):
This looks to me like it is going in the right direction. Some comments:
* Table data fetching with login is fast and smooth. Form saving however seems very slow, it takes about a minute for a login dialog to pop up.
* We shouldn't have to manually call `clearLoginHandler` after every successful fetchWithLogin, this should be done internally in the fetchWithLogin function. Basically as a dev using this component, I should be able to replace `fetch` with `fetchWithLogin` with the same params and not worry about anything else.
* Currently attempts to save a form while logged out populate the JS console with this kind of errors:
```
Uncaught (in promise) 
Response {type: "basic", url: "http://localhost:8655/Forms/c9d4834e-a79e-4268-b61f-2c6d2c884a9a", redirected: false, status: 500, ok: false, …}
```
* When attempting to fetch data for the Subjects table while not logged in this error shows up in the JS console 3 times:
```
Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position 3
```
* After testing this and seeing the login dialog pop up many times, I am now ready to remove the "Register" button from it :)

sashaandjic commented:
Not sure if this is a problem. I loged out in one tab, in the other one I tried to add a filter
![image](https://user-images.githubusercontent.com/23243244/101093906-8b5f4380-3589-11eb-9ce8-128b673fbcee.png)

Also I noticed in console the following error happening quite often:
![image](https://user-images.githubusercontent.com/23243244/101094024-bea1d280-3589-11eb-88c3-e299954a7952.png)

sashaandjic commented:
I had LFS opened in 2 tabs. On one tab I had Dashboard , on the other one I was editing Radiation form. On the tab with Dashboard I logged out, after which I was able to continue with Radiotherapy form edits. Once I clicked on SAVE button, the form seems to get frozen.
![image](https://user-images.githubusercontent.com/23243244/101096298-610f8500-358d-11eb-8728-112f32e1deb6.png)
Then I logged in the first tab (Dashboard), when I refreshed the other tab with Radio form I was prompted to logi in again. Is this expected?

sdumitriu commented on distribution/src/main/provisioning/50-lfs.txt:
```diff
@@ -30,7 +30,8 @@
     # Add support for external authenticators
-    ca.sickkids.ccm.lfs/lfs-ldap-support/${lfs.version}/slingfeature
+    # Uncomment to enable LDAP support
+    #ca.sickkids.ccm.lfs/lfs-ldap-support/${lfs.version}/slingfeature
```
Instead of disabling, this should be behind a runmode.

marta- replied:
I think this should be disabled by default for now, and should be re-enabled and configured for each project as needed upon deployment. When not properly set up, it just causes long delays.

FTR, `lfs` and `cardiac` runmodes will need LDAP, but `cardiac_rehab` will not for now.

marta- reviewed (commented):
So far, so good :+1:

sdumitriu reviewed (commented):
Looks good so far.

sashaandjic commented:
no complaints from me either

sashaandjic reviewed (approved):
tested
I didn't run into any problem.

IntegralProgrammer committed:
LFS-655: Handle re-login globally

Create the global React context for the re-login dialog

IntegralProgrammer committed:
LFS-655: Handle re-login globally

Replace fetch() with fetchWithReLogin() in LiveTable.jsx

IntegralProgrammer committed:
LFS-655: Handle re-login globally

Replace fetch() with fetchWithReLogin() in Form.jsx

IntegralProgrammer committed:
LFS-655: Handle re-login globally

Replace fetch() with fetchWithReLogin() in Subject.jsx

IntegralProgrammer committed:
LFS-655: Handle re-login globally

Replace fetch() with fetchWithReLogin() in SubjectView.jsx and UserDashboard.jsx

IntegralProgrammer committed:
LFS-655: Handle re-login globally

Replace fetch() with fetchWithReLogin() in Filters.jsx

IntegralProgrammer committed:
LFS-655: Handle re-login globally

Replace fetch() with fetchWithReLogin() in SearchBar.jsx

IntegralProgrammer committed:
LFS-655: Handle re-login globally

Replace fetch() with fetchWithReLogin() in NewFormDialog.jsx and SubjectSelector.jsx

marta- committed:
Merge branch 'dev' into LFS-655

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -332,7 +318,7 @@ function Form (props) {
             />
           </Typography>
-          <Breadcrumbs separator="·">
+          <Breadcrumbs separator=".">
```
This needs to be reverted
```suggestion
          <Breadcrumbs separator="·">
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -310,10 +295,11 @@ function Form (props) {
       </Grid>
     );
   }
+
```
```suggestion
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -647,7 +657,7 @@ export const SelectorDialog = withStyles(QuestionnaireStyle)(UnstyledSelectorDia
  * @param {func} onError The callback if an error occurs during subject creation
  */
-export function createSubjects(newSubjects, subjectType, subjectParents, subjectToTrack, returnCall, onError) {
+export function createSubjects(globalLoginDisplay, newSubjects, subjectType, subjectParents, subjectToTrack, returnCall, onError) {
```
The new parameter needs to be documented

sdumitriu committed:
[misc] Undo bad change

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

sdumitriu reviewed (approved):
The usual workflow seems to work very well.

Things that are still not handled correctly:
- the administration section appears to be largely accessible, even though actually changing things doesn't work
- while the Subjects page prompts for login, the Patients and Tumors still display the JSON parse syntax error
- the Variants upload page shows `undefined undefined` in the file selector button
